add test_get_orders, test_create_order, add test_create_order, test_create_user, test_update_user_data

Zorkiy87 opened the pull request:
add test_create_order.py
add test_login_user.py
add test_create_user.py
add test_update_user_data.py

Zorkiy87 committed:
add test_get_orders.py
add test_create_order.py
add test_login_user.py
add test_create_user.py
add test_update_user_data.py

DolgoarshinnykhLG commented on data.py:
```diff
@@ -0,0 +1,16 @@
+
+
+class Urls:
+    base_url = 'https://stellarburgers.nomoreparties.site'
+    url_create_user = f'{base_url}/api/auth/register'
+    url_user_login = f'{base_url}/api/auth/login'
+    url_delete_user = f'{base_url}/api/auth/user'
+    url_update_user = f'{base_url}/api/auth/user'
+    url_create_order = f'{base_url}/api/orders'
+    url_get_order = f'{base_url}/api/orders'
```
Можно лучше: адреса лучше вынести во внешний модуль(например, urls). Они могут измениться, что затруднит поддержку тестов

DolgoarshinnykhLG commented on tests/test_create_order.py:
```diff
@@ -0,0 +1,58 @@
+        headers = {
+            'Authorization': f'{access_token}',
+        }
+        new_response = requests.post(Urls.url_create_order, json=new_payload, headers=headers )
```
Необходимо исправить здесь и далее: необходимо сделать allure-аннотации для каждого метода отправки запроса. При такой реализации лучше всего подходит контекст-менеджер with allure.step

Zorkiy87 replied:
Исправлено.

DolgoarshinnykhLG commented on tests/test_create_user.py:
```diff
@@ -0,0 +1,69 @@
+
+    @allure.title('Создаем нового пользователя, через рандом с уникальными данными')
+    def test_create_new_user(self, registration_and_delete_user):
+        payload, response = registration_and_delete_user
```
Необходимо исправить: создание пользователя в этом сценарии - это шаг теста, а сейчас он фактически выполняется в фикстуре. Для этого теста необходимо отправку запроса на создание пользователя включить в шаги теста

Zorkiy87 replied:
Исправлено.

DolgoarshinnykhLG commented on tests/test_create_user.py:
```diff
@@ -0,0 +1,69 @@
+    @allure.title('Создаем нового пользователя, через рандом с уникальными данными')
+    def test_create_new_user(self, registration_and_delete_user):
+        payload, response = registration_and_delete_user
+        assert response.status_code == 200, f"Ожидался статус 200, но получен {response.status_code}"
```
Необходимо исправитНужно исправить здесь и далее: в этих сценариях недостаточно проверять только код ответа. В ответ нам возвращается, как правило, и тело ответа. Необходимо проверить и его

Zorkiy87 replied:
Исправлено.

DolgoarshinnykhLG commented on tests/test_create_user.py:
```diff
@@ -0,0 +1,69 @@
+            f"Ожидалось сообщение'Email, password and name are required fields', но получен ответ {response.json()['message']}"
+        assert response.json()['success'] == False, \
+            f"Ожидался ответ False, но получен ответ {response.json()['success']}"
+
```
Можно лучше: шаги этих тестов схожи. Можно их объединить с помощью параметризации

Zorkiy87 committed:
1) Внесены allure аннотации через with allure.step
2) In tests/test_create_user.py: переработаны тесты. Пользователь создается в шаге теста, а не в фикстуре.
3) In tests/test_create_user.py: добавлены проверки для тела ответа.

DolgoarshinnykhLG commented on tests/test_create_user.py:
```diff
@@ -0,0 +1,98 @@
+            assert response.json()['success'] == True
+            assert response.json()['user']['email'] == payload['email']
+            assert response.json()['user']['name'] == payload['name']
+        with allure.step('Получаем ключ доступа для запроса'):
+            access_token = response.json().get('accessToken')
+        with allure.step('Отправляем запрос на удаление созданного пользователя'):
+            requests.delete(Urls.url_delete_user, headers={'Authorization': access_token})
```
Необходимо исправить здесь и далее: удаление пользователя - это часть пост-условия, а не шаг теста. Лучше его вынести в фикстуры

Zorkiy87 replied:
Не согласен.
1) У меня реализовано создание и удаление пользователя через фикстуру, в предыдущем ревью надо было вынести создание пользователя из фикстуры в шаги теста, что я и сделал.
2) Создание пользователя и удаление в шагах теста используется в 2 тестах test_create_new_user и test_cant_create_two_identical_users, в остальных тестах (14), реализовано через фикстуру создания и удаления пользователя.
3) Считаю делать фикстуру удаления пользователя из-за двух тестов нецелесообразно. Если это реализовывать, то надо добавлять функцию передачи параметров созданного пользователя в фикстуру для удаления, но это только увеличит количество кода, усложнит логику и придётся переписывать фикстуру создания и удаления пользователя, так как добавление второй фикстуры приведёт к дублированию функций.

DolgoarshinnykhLG replied:
В тестах не должно быть шагов, которые к телу теста не относятся. Это либо пост-условие, либо предусловия к нему. В этом кейсе создание - целевой шаг теста, в фикстуру его вынести нельзя (кроме тестов с повторным созданием пользователя и всеми остальным, там это явное предусловие), но можно написать фикстуру по такой логике: генерируем данные через generate_random_email_password_name и возвращаем в тест, а в пост-условиях пытаемся провести удаление. Так мы сделаем тест чистым и почистим тестовые данные
